fedimint-api/src/multiplexed.rs

```diff
@@ -0,0 +1,180 @@
+        Self {
+            inner: Arc::new(ModuleMultiplexerInner {
+                connections: Mutex::new(connections),
+                out_of_order: Default::default(),
+            }),
+        }
+    }
+
+    #[cfg(target_family = "wasm")]
+    pub fn new(connections: AnyPeerConnections<ModuleMultiplexed<Msg>>) -> Self {
+        unimplemented!();
+    }
+
+    #[cfg(not(target_family = "wasm"))]
+    pub async fn send(&self, peers: &[PeerId], module: &str, msg: Msg) -> Cancellable<()> {
+        debug!("Sending to {peers:?}: to {module}, {msg:?}");
+        self.inner
+            .connections
+            .lock()
+            .await
+            .send(
+                peers,
+                ModuleMultiplexed {
+                    module: module.into(),
+                    msg,
+                },
+            )
+            .await
+    }
+
+    #[cfg(target_family = "wasm")]
+    pub async fn send(&self, peers: &[PeerId], module: &str, msg: Msg) -> Cancellable<()> {
+        unimplemented!();
+    }
+
+    #[cfg(not(target_family = "wasm"))]
+    /// Await receipt of a message from any connected peer.
+    pub async fn receive(&self, module: &str) -> Cancellable<(PeerId, Msg)> {
+        loop {
+            // Note: tokio locks are FIFO, so no need to sleep between loop iterations,
+            // as each thread should get a chance to check for a message.
+            // However, there is a one minor problem with this implementation -
+            // other threads can only check if there is something for them,
+            // when
+            // Note: Naive implementation could have some subtle problems. If
+            // one lock was used, either other threads could not check `out_of_order`
+            // while some thread is blocked (and holding the lock) in `recv`,
+            // or there would be subtle errors where a thread could miss a newly
+            // received item and go into `recv`, possibly blocking indefinitely
+            // if no more messages are being delivered.
+            let mut out_of_order = self.inner.out_of_order.lock().await;
+            if let Some(queue) = out_of_order.msgs.get_mut(module) {
+                if let Some(existing) = queue.pop_front() {
+                    *out_of_order
+                        .peer_counts
+                        .get_mut(&existing.0)
+                        .expect("peer must have an entry if had a message already") -= 1;
+                    return Ok(existing);
+                }
+            }
+
+            // try lock is used to avoid a deadlock (see below)
+            if let Ok(mut connections) = self.inner.connections.try_lock() {
+                // `out_of_order` lock guard is dropped *after* we obtained `connections`,
+                // guaranteeing that new elements could have been added to `out_of_order`
+                // since we've last checked.
+                drop(out_of_order);
```

Comment:
Fortunately multiplexing is only used for DKG, otherwise I'd really want some stress/fuzz tests here to avoid deadlocks that would be nearly impossible to debug imo.